File: go/ai/llm/router/controller_test.go
package router

import (
	"context"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	"hop.top/kit/go/ai/llm"
)

func TestParseModelString(t *testing.T) {
	tests := []struct {
		name      string
		model     string
		wantName  string
		wantThres float64
		wantErr   bool
	}{
		{
			name:      "valid mf",
			model:     "router-mf-0.116",
			wantName:  "mf",
			wantThres: 0.116,
		},
		{
			name:      "valid bert",
			model:     "router-bert-0.7",
			wantName:  "bert",
			wantThres: 0.7,
		},
		{
			name:      "valid sw_ranking",
			model:     "router-sw_ranking-0.5",
			wantName:  "sw_ranking",
			wantThres: 0.5,
		},
		{
			name:      "threshold zero",
			model:     "router-random-0",
			wantName:  "random",
			wantThres: 0,
		},
		{
			name:      "threshold one",
			model:     "router-random-1",
			wantName:  "random",
			wantThres: 1,
		},
		{
			name:    "missing prefix",
			model:   "mf-0.5",
			wantErr: true,
		},
		{
			name:    "no threshold",
			model:   "router-mf",
			wantErr: true,
		},
		{
			name:    "invalid threshold",
			model:   "router-mf-abc",
			wantErr: true,
		},
		{
			name:    "threshold out of range high",
			model:   "router-mf-1.5",
			wantErr: true,
		},
		{
			name:      "hyphenated router name",
			model:     "router-sw-ranking-0.5",
			wantName:  "sw-ranking",
			wantThres: 0.5,
		},
		{
			name:    "empty string",
			model:   "",
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			name, threshold, err := ParseModelString(tt.model)
			if tt.wantErr {
				assert.Error(t, err)
				return
			}
			require.NoError(t, err)
			assert.Equal(t, tt.wantName, name)
			assert.InDelta(t, tt.wantThres, threshold, 0.0001)
		})
	}
}

func TestController_Route(t *testing.T) {
	reg := NewRegistry()
	_ = reg.Register("test", &stubRouter{score: 0.8})

	ctrl := NewController(reg, ModelPair{
		Strong: "gpt-4",
		Weak:   "gpt-3.5",
	})

	ctx := context.Background()
	sig := UserSignal{Text: "hello"}

	// Score 0.8 >= threshold 0.5 => strong
	model, err := ctrl.Route(ctx, sig, "test", 0.5)
	require.NoError(t, err)
	assert.Equal(t, "gpt-4", model)

	// Score 0.8 < threshold 0.9 => weak
	model, err = ctrl.Route(ctx, sig, "test", 0.9)
	require.NoError(t, err)
	assert.Equal(t, "gpt-3.5", model)
}

func TestController_RouteFromModel(t *testing.T) {
	reg := NewRegistry()
	_ = reg.Register("mf", &stubRouter{score: 0.3})

	ctrl := NewController(reg, ModelPair{
		Strong: "gpt-4",
		Weak:   "gpt-3.5",
	})

	ctx := context.Background()
	sig := UserSignal{Text: "hello"}

	model, err := ctrl.RouteFromModel(ctx, sig, "router-mf-0.5")
	require.NoError(t, err)
	assert.Equal(t, "gpt-3.5", model) // 0.3 < 0.5 => weak
}

func TestController_Route_InvalidRouter(t *testing.T) {
	reg := NewRegistry()
	ctrl := NewController(reg, ModelPair{Strong: "a", Weak: "b"})

	_, err := ctrl.Route(context.Background(), UserSignal{Text: "hello"}, "nonexistent", 0.5)
	assert.Error(t, err)
}

func TestController_Route_InvalidThreshold(t *testing.T) {
	reg := NewRegistry()
	_ = reg.Register("test", &stubRouter{score: 0.5})
	ctrl := NewController(reg, ModelPair{Strong: "a", Weak: "b"})

	_, err := ctrl.Route(context.Background(), UserSignal{Text: "hello"}, "test", 1.5)
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "out of [0,1] range")
}

// stubMiddleware overrides model pair.
type stubMiddleware struct {
	pair *ModelPair
}

func (s *stubMiddleware) GetModelPair(
	_ context.Context, _ string,
) (*ModelPair, error) {
	return s.pair, nil
}

func TestController_Middleware(t *testing.T) {
	reg := NewRegistry()
	_ = reg.Register("test", &stubRouter{score: 0.8})

	mw := &stubMiddleware{pair: &ModelPair{
		Strong: "claude-3",
		Weak:   "claude-haiku",
	}}

	ctrl := NewController(reg, ModelPair{
		Strong: "gpt-4",
		Weak:   "gpt-3.5",
	}, WithMiddleware(mw))

	ctx := context.Background()

	// Middleware overrides pair; 0.8 >= 0.5 => strong = claude-3
	model, err := ctrl.Route(ctx, UserSignal{Text: "hello"}, "test", 0.5)
	require.NoError(t, err)
	assert.Equal(t, "claude-3", model)
}

func TestController_ModelCounts(t *testing.T) {
	reg := NewRegistry()
	_ = reg.Register("test", &stubRouter{score: 0.8})

	ctrl := NewController(reg, ModelPair{
		Strong: "gpt-4",
		Weak:   "gpt-3.5",
	})

	ctx := context.Background()
	_, _ = ctrl.Route(ctx, UserSignal{Text: "a"}, "test", 0.5) // strong
	_, _ = ctrl.Route(ctx, UserSignal{Text: "b"}, "test", 0.5) // strong
	_, _ = ctrl.Route(ctx, UserSignal{Text: "c"}, "test", 0.9) // weak

	counts := ctrl.ModelCounts()
	assert.Equal(t, 2, counts["test"]["gpt-4"])
	assert.Equal(t, 1, counts["test"]["gpt-3.5"])
}

func TestController_Complete_NoProvider(t *testing.T) {
	reg := NewRegistry()
	_ = reg.Register("test", &stubRouter{score: 0.5})

	ctrl := NewController(reg, ModelPair{Strong: "a", Weak: "b"})

	_, err := ctrl.Complete(context.Background(), llm.Request{
		Model:    "router-test-0.5",
		Messages: []llm.Message{{Role: "user", Content: "hi"}},
	})
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "no provider configured")
}

func TestLastUserSignal(t *testing.T) {
	t.Run("text only", func(t *testing.T) {
		msgs := []llm.Message{
			{Role: "system", Content: "you are helpful"},
			{Role: "user", Content: "first question"},
			{Role: "assistant", Content: "first answer"},
			{Role: "user", Content: "second question"},
		}
		sig := lastUserSignal(msgs)
		assert.Equal(t, "second question", sig.Text)
		assert.False(t, sig.HasImage)
		assert.False(t, sig.HasAudio)
		assert.False(t, sig.HasVideo)
	})

	t.Run("no user message falls back to last", func(t *testing.T) {
		msgs := []llm.Message{
			{Role: "system", Content: "system prompt"},
		}
		sig := lastUserSignal(msgs)
		assert.Equal(t, "system prompt", sig.Text)
	})

	t.Run("empty messages", func(t *testing.T) {
		sig := lastUserSignal(nil)
		assert.Equal(t, "", sig.Text)
	})

	t.Run("multimodal parts set flags", func(t *testing.T) {
		msgs := []llm.Message{
			{Role: "user", Parts: []llm.ContentPart{
				{Type: llm.PartTypeText, Text: "describe this"},
				{Type: llm.PartTypeImage},
			}},
		}
		sig := lastUserSignal(msgs)
		assert.Equal(t, "describe this", sig.Text)
		assert.True(t, sig.HasImage)
		assert.False(t, sig.HasAudio)
		assert.False(t, sig.HasVideo)
	})

	t.Run("audio and video flags", func(t *testing.T) {
		msgs := []llm.Message{
			{Role: "user", Parts: []llm.ContentPart{
				{Type: llm.PartTypeAudio},
				{Type: llm.PartTypeVideo},
			}},
		}
		sig := lastUserSignal(msgs)
		assert.True(t, sig.HasAudio)
		assert.True(t, sig.HasVideo)
		assert.False(t, sig.HasImage)
	})

	t.Run("parts text concatenated", func(t *testing.T) {
		msgs := []llm.Message{
			{Role: "user", Parts: []llm.ContentPart{
				{Type: llm.PartTypeText, Text: "hello"},
				{Type: llm.PartTypeImage},
				{Type: llm.PartTypeText, Text: " world"},
			}},
		}
		sig := lastUserSignal(msgs)
		assert.Equal(t, "hello world", sig.Text)
	})

	t.Run("Content used when Parts empty", func(t *testing.T) {
		msgs := []llm.Message{
			{Role: "user", Content: "plain text", Parts: nil},
		}
		sig := lastUserSignal(msgs)
		assert.Equal(t, "plain text", sig.Text)
		assert.False(t, sig.HasImage)
	})
}

func TestController_Route_ModalRouter(t *testing.T) {
	reg := NewRegistry()
	_ = reg.Register("modal", &stubModalRouter{score: 0.9})

	ctrl := NewController(reg, ModelPair{
		Strong: "gpt-4-vision",
		Weak:   "gpt-3.5",
	})

	sig := UserSignal{Text: "describe", HasImage: true}
	model, err := ctrl.Route(context.Background(), sig, "modal", 0.5)
	require.NoError(t, err)
	assert.Equal(t, "gpt-4-vision", model)
}

// stubModalRouter implements ModalRouter.
type stubModalRouter struct {
	score float64
}

func (s *stubModalRouter) Score(_ context.Context, _ string) (float64, error) {
	return 0.1, nil // low score via text-only path
}

func (s *stubModalRouter) ScoreSignal(_ context.Context, _ UserSignal) (float64, error) {
	return s.score, nil
}

File: sdk/ts/src/router/controller.test.ts
import { describe, it, expect, vi } from "vitest";
import { Controller, parseModelName } from "./controller";
import { RoutingError } from "./router";
import type { Router, Middleware, ModelPair } from "./router";

// ─── Helpers ─────────────────────────────────────────────────────────────────

function fixedRouter(value: number): Router {
  return { score: vi.fn(async () => value) };
}

function overrideMiddleware(pair: ModelPair): Middleware {
  return { getModelPair: vi.fn(async () => pair) };
}

function nullMiddleware(): Middleware {
  return { getModelPair: vi.fn(async () => null) };
}

// ─── parseModelName ──────────────────────────────────────────────────────────

describe("parseModelName", () => {
  it("parses router-mf-0.7", () => {
    const r = parseModelName("router-mf-0.7");
    expect(r.routerName).toBe("mf");
    expect(r.threshold).toBe(0.7);
  });

  it("parses router-sw_ranking-0.5", () => {
    const r = parseModelName("router-sw_ranking-0.5");
    expect(r.routerName).toBe("sw_ranking");
    expect(r.threshold).toBe(0.5);
  });

  it("parses threshold 0 and 1", () => {
    expect(parseModelName("router-bert-0").threshold).toBe(0);
    expect(parseModelName("router-bert-1").threshold).toBe(1);
  });

  it("throws when prefix is missing", () => {
    expect(() => parseModelName("mf-0.5")).toThrow(RoutingError);
  });

  it("throws when threshold is missing", () => {
    expect(() => parseModelName("router-mf")).toThrow(RoutingError);
  });

  it("throws on non-numeric threshold", () => {
    expect(() => parseModelName("router-mf-abc")).toThrow(RoutingError);
  });
});

// ─── Controller ──────────────────────────────────────────────────────────────

describe("Controller", () => {
  it("routes to strong model when score >= threshold", async () => {
    const ctrl = new Controller({
      strongModel: "gpt-4",
      weakModel: "gpt-3.5",
      routers: { mf: fixedRouter(0.8) },
    });

    const model = await ctrl.route("hello", "mf", 0.5);
    expect(model).toBe("gpt-4");
  });

  it("routes to weak model when score < threshold", async () => {
    const ctrl = new Controller({
      strongModel: "gpt-4",
      weakModel: "gpt-3.5",
      routers: { mf: fixedRouter(0.3) },
    });

    const model = await ctrl.route("hello", "mf", 0.5);
    expect(model).toBe("gpt-3.5");
  });

  it("throws on unknown router", async () => {
    const ctrl = new Controller({
      strongModel: "gpt-4",
      weakModel: "gpt-3.5",
    });

    await expect(ctrl.route("hi", "missing", 0.5)).rejects.toThrow(
      RoutingError,
    );
  });

  it("throws on invalid threshold", async () => {
    const ctrl = new Controller({
      strongModel: "gpt-4",
      weakModel: "gpt-3.5",
      routers: { mf: fixedRouter(0.5) },
    });

    await expect(ctrl.route("hi", "mf", 1.5)).rejects.toThrow(
      RoutingError,
    );
    await expect(ctrl.route("hi", "mf", -0.1)).rejects.toThrow(
      RoutingError,
    );
  });

  it("tracks model counts", async () => {
    const ctrl = new Controller({
      strongModel: "gpt-4",
      weakModel: "gpt-3.5",
      routers: { mf: fixedRouter(0.8) },
    });

    await ctrl.route("a", "mf", 0.5);
    await ctrl.route("b", "mf", 0.5);
    const counts = ctrl.modelCounts.get("mf")!;
    expect(counts.get("gpt-4")).toBe(2);
  });

  it("addRouter registers a new router", async () => {
    const ctrl = new Controller({
      strongModel: "gpt-4",
      weakModel: "gpt-3.5",
    });
    ctrl.addRouter("test", fixedRouter(1));
    expect(ctrl.routerNames()).toContain("test");
    const m = await ctrl.route("hi", "test", 0.5);
    expect(m).toBe("gpt-4");
  });

  it("routeByModelName delegates correctly", async () => {
    const ctrl = new Controller({
      strongModel: "gpt-4",
      weakModel: "gpt-3.5",
      routers: { bert: fixedRouter(0.9) },
    });

    const m = await ctrl.routeByModelName("hi", "router-bert-0.5");
    expect(m).toBe("gpt-4");
  });

  it("middleware can override model pair", async () => {
    const custom: ModelPair = {
      strong: "claude-3",
      weak: "mixtral",
    };
    const ctrl = new Controller({
      strongModel: "gpt-4",
      weakModel: "gpt-3.5",
      routers: { mf: fixedRouter(0.9) },
      middleware: [overrideMiddleware(custom)],
    });

    const m = await ctrl.route("hi", "mf", 0.5);
    expect(m).toBe("claude-3");
  });

  it("null-returning middleware keeps default pair", async () => {
    const ctrl = new Controller({
      strongModel: "gpt-4",
      weakModel: "gpt-3.5",
      routers: { mf: fixedRouter(0.9) },
      middleware: [nullMiddleware()],
    });

    const m = await ctrl.route("hi", "mf", 0.5);
    expect(m).toBe("gpt-4");
  });

  it("last middleware wins", async () => {
    const first: ModelPair = { strong: "a", weak: "b" };
    const second: ModelPair = { strong: "x", weak: "y" };
    const ctrl = new Controller({
      strongModel: "gpt-4",
      weakModel: "gpt-3.5",
      routers: { mf: fixedRouter(0.1) },
      middleware: [
        overrideMiddleware(first),
        overrideMiddleware(second),
      ],
    });

    const m = await ctrl.route("hi", "mf", 0.5);
    expect(m).toBe("y"); // weak, score=0.1 < 0.5
  });
});
